pkg/kubectl/rollback.go

```diff
@@ -0,0 +1,118 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubectl
+
+import (
+	"fmt"
+	"os"
+	"os/signal"
+	"syscall"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/runtime"
+	deploymentutil "k8s.io/kubernetes/pkg/util/deployment"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+// Rollbacker provides an interface for resources that can be rolled back.
+type Rollbacker interface {
+	Rollback(namespace, name string, updatedAnnotations map[string]string, toRevision int64, obj runtime.Object) (string, error)
+}
+
+func RollbackerFor(kind unversioned.GroupKind, c client.Interface) (Rollbacker, error) {
+	switch kind {
+	case extensions.Kind("Deployment"):
+		return &DeploymentRollbacker{c}, nil
+	}
+	return nil, fmt.Errorf("no rollbacker has been implemented for %q", kind)
+}
+
+type DeploymentRollbacker struct {
+	c client.Interface
+}
+
+func (r *DeploymentRollbacker) Rollback(namespace, name string, updatedAnnotations map[string]string, toRevision int64, obj runtime.Object) (string, error) {
+	deploymentRollback := &extensions.DeploymentRollback{
+		Name:               name,
+		UpdatedAnnotations: updatedAnnotations,
+		RollbackTo: extensions.RollbackConfig{
+			Revision: toRevision,
+		},
+	}
+	result := ""
+
+	// Get current events
+	events, err := r.c.Events(namespace).List(api.ListOptions{})
+	if err != nil {
+		return result, err
+	}
+	// Do the rollback
+	if err := r.c.Extensions().Deployments(namespace).Rollback(deploymentRollback); err != nil {
+		return result, err
+	}
+	// Watch for the changes of events
+	watch, err := r.c.Events(namespace).Watch(api.ListOptions{Watch: true, ResourceVersion: events.ResourceVersion})
```

Comment:
If we don't list and get `ResourceRevision` to set in Watch, Watch will return all events back, not just the latest ones. (i.e. we don't know what's updated after the call to Rollback)